File: day-02/sol.py
#!/usr/bin/env python3

import typing as t


Instructions = t.Sequence[t.Tuple[str, int]]


def parse_input(input: str) -> Instructions:
    instructions = []
    for line in input.splitlines():
        op, val = line.split(sep=None, maxsplit=1)
        instructions.append((op, int(val)))
    return instructions


def part1(instructions: Instructions) -> int:
    hpos = 0
    depth = 0
    for (op, val) in instructions:
        if op == "forward":
            hpos += val
        else:
            val *= (op == "down") * 2 - 1
            depth += val
    return hpos * depth


def part2(instructions: Instructions) -> int:
    hpos = 0
    depth = 0
    aim = 0
    for (op, val) in instructions:
        if op == "forward":
            hpos += val
            depth += aim * val
        else:
            val *= (op == "down") * 2 - 1
            aim += val
    return hpos * depth


example_input = parse_input("""\
forward 5
down 5
forward 8
up 3
down 8
forward 2""")
print(part1(example_input), 150)
print(part2(example_input), 900)

my_input = parse_input(open("input.txt").read())
print(part1(my_input), 1561344)
print(part2(my_input), 1848454425)

File: day-02/sol.clj
#!/usr/bin/env bb

(ns day2
  (:require [clojure.string :as str]))


(defn parse-input [input]
  (->> input
       str/split-lines
       (map #(str/split % #"\s"))
       (map (fn [[cmd n]] [cmd (Integer/parseInt n)]))))


(def example-input (parse-input "forward 5
down 5
forward 8
up 3
down 8
forward 2
"))

(def my-input (parse-input (slurp "input.txt")))

(defn part1 [input]
  (->> input
       (reduce
         (fn [[hpos depth] [cmd n]]
           (case cmd
             "up" [hpos (- depth n)]
             "down" [hpos (+ depth n)]
             "forward" [(+ hpos n) depth]))
        [0 0])
       (apply *)))

(defn part2 [input]
  (->> input
       (reduce
         (fn [[hpos depth aim] [cmd n]]
           (case cmd
             "up" [hpos depth (- aim n)]
             "down" [hpos depth (+ aim n)]
             "forward" [(+ hpos n) (+ depth (* aim n)) aim]))
        [0 0 0])
       butlast
       (apply *)))


(prn (part1 example-input) 150)
(prn (part2 example-input) 900)
(prn (part1 my-input) 1561344)
(prn (part2 my-input) 1848454425)
